libobs-d3d11/d3d11-shader.cpp

```diff
@@ -200,26 +201,71 @@ void gs_shader::BuildConstantBuffer()
 		gs_shader_set_default(&params[i]);
 }
 
+static uint64_t fnv1a_hash(const char *str)
+{
+	const uint64_t FNV_OFFSET = 14695981039346656037ULL;
+	const uint64_t FNV_PRIME = 1099511628211ULL;
+	uint64_t hash = FNV_OFFSET;
+	while (*str) {
+		hash ^= (uint64_t)*str++;
+		hash *= FNV_PRIME;
+	}
+	return hash;
+}
+
 void gs_shader::Compile(const char *shaderString, const char *file,
 			const char *target, ID3D10Blob **shader)
 {
 	ComPtr<ID3D10Blob> errorsBlob;
 	HRESULT hr;
 
+	FILE *cache_file;
+	struct dstr cache_path = {0};
+	char hashstr[20];
+
 	if (!shaderString)
 		throw "No shader string specified";
 
-	hr = device->d3dCompile(shaderString, strlen(shaderString), file, NULL,
-				NULL, "main", target,
-				D3D10_SHADER_OPTIMIZATION_LEVEL1, 0, shader,
-				errorsBlob.Assign());
-	if (FAILED(hr)) {
-		if (errorsBlob != NULL && errorsBlob->GetBufferSize())
-			throw ShaderError(errorsBlob, hr);
-		else
-			throw HRError("Failed to compile shader", hr);
+	uint64_t hash = fnv1a_hash(shaderString);
+	snprintf(hashstr, sizeof(hashstr), "%02llx", hash);
+
+	dstr_init_move_array(&cache_path, os_get_program_data_path_ptr(
+						  "obs-studio/shader-cache"));
+	os_mkdirs(cache_path.array);
```

Comment:
This should ideally be done somewhere else, but where?

Reply:
Wheh running in portable mode, would this not still use appdata for the cache?

Reply:
I suppose this could be done in `MakeUserDirs()` in `obs-app.cpp`, that should run before libobs-d3d11 is loaded. If this gets merged we should probably also add the clearing/removal of that folder to the NSIS/Steam install script like we do with `%PROGRAMDATA%\obs-studio-hook`.

Reply:
I had considered that but thought it was bad design since there's no guarantee that libobs-d3d11 is the graphics backend that ends up being used. I suppose for now as it being the only supported one it's fine.

Reply:
That is true, if we get back into the whole libobs vs UI thing then relying on the UI creating the folder may be considered bad practice as well, though in praxis it doesn't matter for us. Maybe there's a better place withing libobs-d3d11 itself during startup.

Reply:
What about doing it in the `device_create` function like the logging of adapaters? https://github.com/derrod/obs-studio/commit/114bd0aa573f830a2c2644d5768497c3ab733a45